### DISCUSSION TITLE
I would like to ask for some suggestion about using OxyPlot

### DISCUSSION THREAD

**Discussion by newxiumu** [oxyplot/oxyplot#1791] [Q&A]
Hello, maintainer

I'm here again.

I would like to ask for some suggestion, because it has been bothering me for a long time. Can I use Oxyplot to process about 20-30 plots on a view and refresh the data every 500ms? there are about 10000 data on each plot.

In the process of writing my own code, I found that when the data is a little larger, it gets stuck. I don't know if there is a problem with my code or the performance of plot is not suitable?

Thank you very much. I like your project very much.

.Net5 OxyPlot 2.0 WPF

**Comment by VisualMelon** [comment]
How do you mean 'it gets stuck'?

It's hard to comment on performance without seeing the specifics of your models and implementation. There is some discussion about refreshing plots and how it can go wrong in #1786; if you are queueing refreshes more quickly than they can be serviced, you will notice a significant reduction in responsiveness.

Performance depends heavily on the platform and the plots themselves (e.g. a million data-points should be cheap if they are pretty much in a straight line or most of them are outside of the plot bounds, but if it's random X/Y noise will quickly become a problem; markers can be very expensive).

If you have large data-sets, then it may be worth considering some sort of pre-processing to reduce the number of points rendered.

**Reply by newxiumu** [reply]
his is a plot in listbox

```XML
<oxy:Plot Grid.Column="4">
                                <oxy:Plot.Axes>
                                    <oxy:LinearAxis Position="Bottom" IsAxisVisible ="False" IsPanEnabled="False" IsZoomEnabled="False"/>
                                    <oxy:LinearAxis Position="Left" IsAxisVisible="False" IsPanEnabled="False" IsZoomEnabled="False"/>
                                </oxy:Plot.Axes>
                                <oxy:Plot.Series>
                                    <oxy:LineSeries ItemsSource="{Binding DataPoints}" Color="Blue" StrokeThickness="1"/>
                                </oxy:Plot.Series>
                            </oxy:Plot>
```

```C#
Task.Run(() =>
               {
                   while (true)
                   {
                       System.Windows.Application.Current.Dispatcher.Invoke((Action)(() => 
                       {
                                   if (afterFilerData != null)
                                   {
                                       for (int i = 0; i < afterFilerData.Length; i++)
                                       {
                                           this.FliterDataPoints.Add(new DataPoint(i, afterFilerData[i]));
                                       }
                                   }
                  } ));
                       Thread.Sleep(500);
                   }
               });
```
FilterDatapoints is a ObservableCollection，and it is the Itemsource of the listbox。afterFilerData is a 10000 length data.And the listbox has 20-30 item。

**Reply by VisualMelon** [reply]
(Note: this concerns the provision now in oxyplot-contrib)

It's not clear how you application works; I'm presuming you must be clearing `afterFilerData` at some point, but because there is no synchronisation apparent, it looks like you will be re-add the same data-point every iteration.

As a rule, you shouldn't use `Thread.Sleep` or `Dispatcher.Invoke` (unless necessary) because they will lock up a thread and that makes the thread pool unhappy. It would be better to use a dispatch timer or an async `Task.Delay` to rate-limit refreshes, and to queue the action asynchronously (e.g. with `InvokeAsync`).

The [`Invoke(Action)`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.invoke?view=net-5.0#System_Windows_Threading_Dispatcher_Invoke_System_Action_) overload uses the highest priority to dispatch the action (e.g. it will occur in preference to rendering): if you have multiple of these loops running, I would expect that to cause the UI to seize. Consider using an overload that allows you to specify a lower priority (e.g. [`Background`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcherpriority?view=net-5.0)).

**Reply by newxiumu** [reply]
Thanks。
Could you give me some example。

I want to monitor the continuous data of some sensors for my experiment in school. I want to display the data of 20-30 sensors in real time on one page, with the acquisition speed of 100Hz, but I choose 500ms-1000ms to refresh. The maximum data length in each plot is 10000. But every time I update the UI, I get stuck.

Thank you very much.
